### ISSUE TITLE
Using hooks within selectors

### ISSUE THREAD

**Issue by grod220** [facebookexperimental/Recoil#1382]
At the moment, selectors do not allow for any hooks inside of it. This is unfortunate because most external libraries provide hooks. This means it's quite difficult to fully move your logic into recoil. You have to rely upon boilerplate patterns like creating custom hooks that sync the result of the hook with recoil atoms.

Here's an example of this common pattern: https://github.com/facebookexperimental/Recoil/issues/727#issuecomment-754456755

Anything on the roadmap related to this?

**Comment by drarmstr** [comment]
Yes, actually.  The synchronization library we're working on has a pattern which helps with syncing atoms with values from hooks.  See [this example](https://github.com/facebookexperimental/Recoil/pull/1252/commits/f1484c9f11b6eea08d28807fb4690468741a517d)  This particular example initializes with data from props, but data from hooks could be used as well as bi-directional syncing and subscriptions.

**Comment by grod220** [comment]
Awesome to hear you are thinking about this! Just some feedback. 

```
      <SyncWithProps spam="SPAM" eggs="EGGS" />
      <ReadsAtom atom={atomA} />
```

I find this pattern to be a bit odd as it moves state logic into a place where UI components are typically expected to be. For example, I think it would be nicer to have something like this:

```
const value = useRecoilValueAndSync(definedAtom, useCustomHook())
```

The best would be being able to use a hook _within_ a selector, but I'd understand if there are inherent limitations with that.

**Comment by drarmstr** [comment]
The intent is to keep the atom-specific sync logic in one place with the atom definition, that's what described what we sync.  It is then abstracted with how it is synced with a hook that acts as a bridge between the non-React atoms and React hooks.  It shouldn't be interlaced with UI logic, that was just a trivial test example.  Hopefully it will be more clear when we get documentation.

**Comment by olefrank** [comment]
I'm looking for common patterns on this. All examples I can find [https://recoiljs.org/docs/guides/asynchronous-data-queries/](https://recoiljs.org/docs/guides/asynchronous-data-queries/) abstracts away the fetch queries... In my application I have access to the fetch queries through a hook which returns a "service" object... Is this pattern doable with recoil?

```
export const getLoanProductsSelector = selector<Product[]>({
  key: 'getProductsSelector',
  get: async ({ get }) => {
    // todo
    const service = useService();
    const calculation = get(calculationFormAtom);
    return service.getProducts(calculation); // returns Promise<Product[]>
  },
});
```
